episodes/testing.md

```diff
@@ -0,0 +1,245 @@
+```
+
+Another way to not do these API calls is by using the [requests_mock library](https://pypi.org/project/requests-mock/) to mock requests API calls. This makes you dependent on another library and still does not show you if things work in reality. It's being used by a lot of people, but personally I prefer fewer dependencies and write integration tests for the integration with external systems.  When you mock this it can give you a false sense of security like happened with the Crowdstrike outage in their testing. If you want to use this an example can be found bellow.
+
+```python
+import requests
+import requests_mock
+
+def get_konrad3d_data(url):
+    response = requests.get(url)
+    return response.text
+
+def test_download_latest_data_konrad3d():
+    data = '<html><head><title>Index of /weather/radar/konrad3d/</title></head><body><h1>Index of /weather/radar/konrad3d/</h1><hr><pre><a href="../">../</a><a href="KONRAD3D_20241116T093000.xml">KONRAD3D_20241116T093000.xml</a>                       16-Nov-2024 09:34                3895<a href="KONRAD3D_20241118T092500.xml">KONRAD3D_20241118T092500.xml</a>                       18-Nov-2024 09:30                3938</pre><hr></body></html>'
+    url = 'https://opendata.dwd.de/weather/radar/konrad3d/'
+    with requests_mock.Mocker() as m:
+        m.get(url, text=data)
+        result = get_konrad3d_data(url)
+    assert result == data
+```
+
+# 4 Performance testing of functions
+
+There are moments that the performance of you function might matter a lot. You might not want a single function to ever execute slower than x seconds. To test this you could write tests for the specific functions that should stay fast. How this works is that you run a function x amount of times and the max duration of the function should not be higher than the x seconds. A useful library to help with these types of test in python is [pytest-benchmark](https://pytest-benchmark.readthedocs.io/en/stable/pedantic.html). This library can also be used to check if the performance between versions of the code is improved.
+
+```python
+import time
+def function_to_test(duration=1):
+    time.sleep(duration)
+    return 123
+
+def test_my_function(benchmark):
+    allowed_speed = 1.000002
+    result = benchmark.pedantic(function_to_test, iterations=5)
+    assert benchmark.stats.stats.max < allowed_speed
+
+    assert result == 123
+```
+
+This code can be run with the following command: `pytest -v -s the file_this_is_in.py::test_my_function`. It will run the code 5 times and none of the calls is allowed to be slower than the allowed_speed.
+
+When you write API's you can also have performance requirements. For this another type of tool is used. One of the most used tools for this in python is locust. FOr more information about this tool look you can look at [their documentation](https://docs.locust.io/en/stable/what-is-locust.html).
+
+# 5. Smoke testing to see if your application is still doing its basic functionality
```

Comment:
Maybe we can replace this section by a reference to https://book.the-turing-way.org/reproducible-research/testing/testing-smoketest

Is there anything domain-specific that we could add (examples of smoke tests in a specific domain)?

Reply:
Added example
Add reference